### ISSUE TITLE
Troubles with Links2 and Userland

### ISSUE THREAD

**Issue by compyl118** [HesselM/rpicross_notes#16]
Hello, 
I have had time to contiue on reading your notes.
I am stuck at step 3 of the userland lib crosscompilation.
For some reason it is not able to finish building the "all" target. Here is the output of the make command:
[output.txt](https://github.com/HesselM/rpicross_notes/files/4410997/output.txt)

Before that everything went well, I managed to complete all the previous steps without any trouble except:
- I did not do the RTC part (althought I activated I2C) because I do not have the hardware
- I had some struggle installing links2 (some network issues it seemed, but after an "apt-get update" I tried again and it worked). 
- I get an error when I try to use links2 "could not initialize any graphics driver. Tried the following drivers: x: Can't open display "(null)" fb: could not get VT mode." I am a little bit confused about how this is supposed to work as neither my VM nor the Rpi has a graphical environment. Am I supposed to SSH on the VM from a HOST with graphical environment for links2 to work? 

Hope you can help me out on this one.
Thanks

**Comment by HesselM** [comment]
Hi compyl118,

Currently I'm in the process of following/checking all steps with Ubuntu 18 and the rasbian Buster release and updating the guide where required. Currently I still need to start the i2c pages and hope to check the userland pages in a couple of days (need to spend time also on other projects too ;) ) - so, if you are ok with it, i'll check in with this issue in a couple of days. 

I guess it could be som kind of driver issue as I experiences also minor other changes in the setup process due to changes in both ubuntu as rasbian.

**Comment by HesselM** [comment]
Ah, I overlooked the 'x' message. If you ssh from you HOST to the RPi you should invoke `-x` in the ssh-call (`HOST$ ssh -x .....`) to allow a display driver to initialise. If you ssh from XCS to the RPi, you need to make sure you already connected a display driver to your XCS. You should be able to do this by connecting the drive to the XCS from HOST with `HOST$ ssh -x ...`, followed by a connection to the RPi with `XCS$ ssh -x ...`

**Comment by compyl118** [comment]
Hi,

Thanks for the reply, I managed to make Links2 work. Turns out, on Putty, enabling X11 is not enough, you have to download Xming and specify a display for it to work. Here is the webpage I used: [https://aruljohn.com/info/x11forwarding/](url)

I think you should warn people that the loading of the image takes a little bit of time. At first, I though it wasn't working properly and closed it prematurely.
--> my bad, you did it already ^^

I still remain stuck at step 3 of userland lib crosscompilation, did you have time to take look at the output file I gave in the first message?  

I have tried to use the buildme script embedded in the git folder (why do you not use it?), it produces the same output. 

Thank you

**Comment by compyl118** [comment]
I managed to compile it on the pi directly using the buildme script and it worked (I would say it took a couple of hours though ^^).

**Comment by HesselM** [comment]
I haven't been able to look at it yet, but slowly I'm getting there. 

Awesome that you managed to get it compiled natively. While this is a nice solution, it obviously defies the purpose of this cross-compile guide.. Which should save you time! ;)

I'll keep this issue open to notify you when this issue is encountered and how it is resolved.

**Comment by compyl118** [comment]
Yes I agree. While I was trying to figure out why it wouldn't work I read this page : 
[https://github.com/raspberrypi/userland/blob/master/README.md](url)

I didn't understand wether the 64bit warning was about hardware or software.
Pizero W is an armV61 core (so not 64bits) but the Linux server you mention in these notes is 64bits. Maybe that's a lead?

**Comment by HesselM** [comment]
So, it took some effort to get to the bottom of this. The short update: I found the issue and updated the guide, sync-scripts and toolchain. 

The longer answer:

As explained in the guide, a cross-compilation setup needs a "sysroot" : a directory structure different that the structure of the host, were we can place all binaries and system files of our target setup. 

In the guide "rootfs" is created as a "sysroot", yet the crosscompiler also had a "sysroot" in `tools/arm-bcm2708/arm-rpi-4.9.3-linux-gnueabihf/arm-linux-gnueabihf/sysroot/`. This sysroot too contains some system files for the RPi, but those are dated as the toolchain hasn't been updated for a while. 

What happend is that the compilation of "userland" mixed newer libraries with old: the userland-build files were excepting libraries with features not yet available when the sysroot of the crosscompiler was created, while the newer libraries did exist in "rootfs".  

So the issue boiled down in ensuring that the compiler picked the newer libraries before the old. This required several adjustments in the toolchain and guide. 

The "hellocam" example also needed a fix as raspbian and userland changed some dependencies.

I've updated and pushed the new files. 

If you are still interested in cross compiling the userland files: it should be working now. If you are planning to give it a try: could you let me know if it all worked out for you?

**Comment by compyl118** [comment]
Thank you. I'm curious as to how you you managed to find that? when I got stuck with the error message I had no clue as to how to continue investigating...

I will start over and try your new guide now. I'll let you know the results.

**Comment by HesselM** [comment]
I can't remember the exact path of discovery, but adding the verbose-flag ("-v") to `CMAKE_CXX_FLAGS` and `CMAKE_C_FLAGS` flags and evaluating the CMakeOutput files shed some light on which libraries and paths where invoked during build and linking. 

Then, some assumptions and discoveries:
- I saw a `fntl64`-linking error in my own setup, when linking the new userland, hinting on 64bit support from some libraries
- I never saw a 64-bit issue during creation of the "old" guide (and hence assumed that the old system-files did not include 64-bit support or reference)
- the repo for crosscompiler is several years old
- the new setup used an updated Raspbian
- the new userland-readme hinted on "64 bit support in development", indicating that some (new) libs had to have 64-bit references.

From this I could make an educated guess that the error meant that the system was linking incorrect libraries, quite possibly the old-sysroot libs instead of the libs synced from the rpi into rootfs and that the new libs should be correct (as you where able to compile it locally, and quite possibly, many others should have been able to do that too: otherwise there would be a huge list of issues and errors on the userland-repo)

As the "old" guide didn't use the proper `CMAKE_SYSROOT` setting (I've toyed with it several years ago) and I knew the [crosscompiler](https://github.com/raspberrypi/tools/tree/master/arm-bcm2708/arm-rpi-4.9.3-linux-gnueabihf) had his own set of libraries, I assumed that this was the task to be solved: finding out how to set up the proper paths.

Finding out the proper paths and settings took a lot of trail and error and diving into the depths of internet into cryptic documented options. 

I guess the result is not yet "perfect" but it is definitely a lot better then the old toolchain and does not require creating (and syncing!) a lot of symlinks as noted in for example [this comment](https://github.com/HesselM/rpicross_notes/issues/10#issuecomment-524153612). I'm still a bit annoyed by the "-B" invocation, but I guess I have to accept this as a solution for now.
